docs/docs/how-to/plugins-and-themes/adding-gatsby-image-support.md

```diff
@@ -0,0 +1,206 @@
+   // so the values will depend on your data type.
+  const filename = image.src
+
+  const sourceMetadata = {
+    width: image.width,
+    height: image.height,
+    format: image.mimeType.split("/")[1]
+  }
+
+  // Generating placeholders is optional, but recommended
+  let placeholderURL
+
+  if(options.placeholder === "blurred") {
+    // This would be your own function to generate a low-resolution placeholder
+    placeholderURL: await getBase64Image({ filename })
+  }
+
+  // You could also calculate dominant color, and pass that as `backgroundColor`
+  // gatsby-plugin-sharp includes helpers that you can use to generate a tracedSVG or calculate
+  // the dominant color of a local file, if you don't want to handle it in your plugin
+
+
+  return generateImageData({
+    ...options,
+    // Passing the plugin name allows for better error messages
+    pluginName: `gatsby-source-example`,
+    sourceMetadata,
+    filename,
+    placeholderURL
+    generateImageSource,
+    options,
+  })
+}
+
+```
+
+### Add the resolver
+
+`gatsby-plugin-image/graphql-utils` includes a utility function to help register the resolver inside the Gatsby `createResolvers` API hook. It registers the resolver with all of the base arguments needed to create the image data, such as width, aspect ratio, layout and background color. These are defined with comprehensive descriptions that are visible when your users are buildign queries in GraphiQL. You can pass additional arguments supported by your plugin, for example image options such as quality.
+
+The arguments:
+
+- `resolverFunction`: the resolver function that you created before. It receives the node and the arguments and should return the image data object.
+- `additionalArgs`: an object defining additional args, in the same format used by [Gatsby Type Builders](https://www.gatsbyjs.com/docs/reference/graphql-data-layer/schema-customization/#gatsby-type-builders)
+
+For example, to add a `gatsbyImageData` resolver onto a `ProductImage` node that you have previously defined:
+
+```js:title=gatsby-source-example/gatsby-node.js
+// Note the different import
+import { getGatsbyImageResolver } from "gatsby-plugin-image/graphql-utils"
+
+export function createResolvers({ createResolvers }) {
+  createResolvers({
+    ProductImage: {
+      // loadImageData is your custom resolver, defined elsewhere
+      gatsbyImageData: getGatsbyImageResolver(loadImageData, {
+        quality: "Int",
+      }),
+    },
+  })
+}
+```
+
+## Adding a custom image component
+
+If you have a URL-based image API, you can create a custom image component that wraps `<GatsbyImage />` and displays images generated at runtime. If you have a source plugin, this could accept your native image object, but it could equally just take a base URL and generate the image based on that. This is a good solution for pure image hosts that aren't handling their own CMS data.
+
+There are two steps to create a custom image component:
```

Comment:
Maybe this should say 3 steps?

Or alternately step 3 could say: 

```
3. Create your wrapper component (optional)
```

so these counts aren't off?